### PR TITLE
clobber pylint errors

### PR DESCRIPTION
**Pull request by bqpd**
@galbramc

### CODE REVIEW

**Comment by galbramc** [comment]
I'll work on the errors in interactive

**Comment by bqpd** [comment]
ah, good plan

**Comment by bqpd** [comment]
getting tests

**Comment by galbramc** [comment]
I don't have permissions to push directly to this branch...

**Comment by bqpd** [comment]
@whoburg can you add @galbramc to this branch? (or how should we do this...)

**Comment by galbramc** [comment]
I think you need to give me permissions to push things directly to the repo. The only other way I know is to fork an make another pull request based on my fork.

**Comment by bqpd** [comment]
I don't think I have gitub-hoburg admin privileges, so I'm happy to redo it as a PR based on your fork.

**Comment by galbramc** [comment]
Just grab the files from here for now:

http://web.mit.edu/galbramc/Public/ractor.py
http://web.mit.edu/galbramc/Public/widgets.py

**Comment by galbramc** [comment]
could you also just double check they are still working properly? They don't have any tests and I don't know how they are supposed to work. I mostly just fixed lines too long from strings.

**Comment by galbramc** [comment]
It looks like expopt.py is not something that was written by your group. If its from the mosek people, then we should not include it in the pylint tests. There is no reason for you to have to fix other peoples problems.

**Comment by bqpd** [comment]
Will do! I'll be offline for a bit now, but back this evening or tomorrow morning.

**Comment by bqpd** [comment]
expopt.py is my translation of mosek C code so it's kinda our problems too...

**Comment by galbramc** [comment]
Then it should be included in pylint. Don't exclude it.

**Comment by galbramc** [comment]
I'll see what I can do with the rest of the remaining ones.

**Comment by galbramc** [comment]
Two more. This is all I think I'm comfortable fixing:

http://web.mit.edu/galbramc/Public/build.py
http://web.mit.edu/galbramc/Public/substitution.py

**Comment by galbramc** [comment]
@whoburg, this is ready for review. We squashed most pylint errors, except for the ones in expopt.py which is mostly complaints about variable naming convention. This one warning however is legit, and somewhat unsettling. I'm not sure what the logic is here, but it looks like the code assumes that the iterator in a for loop takes on a particular value after the for loop ends:

```
gpkit/nomials/variables.py:152: [W0631(undefined-loop-variable), ArrayVariable.__new__] Using possibly undefined loop variable 'value_option'
gpkit/nomials/variables.py:154: [W0631(undefined-loop-variable), ArrayVariable.__new__] Using possibly undefined loop variable 'value_option'
```

**Comment by whoburg** [comment]
you guys rock. I'm going to attempt a rebase (to remove the merge commit and squash a couple trivial commits), force push, and then continue reviewing.

**Comment by whoburg** [comment]
I dropped `# pylint: disable=invalid-name` from `expopt.py` because it triggered new pylint warnings in that file according to Jenkins. Figure we won't touch that file in this PR.

File: gpkit/build.py [hunk 1]
```diff
@@ -111,7 +111,8 @@ def look(self):
             log("#   Trying to run mskexpopt...")
             if call("mskexpopt") in (1052, 28):  # 28 for MacOSX
                 return "in system path"
-        except Exception:
+        except:  # pylint: disable=bare-except
+            # exception type varies by operating system
```
Comment (whoburg) [inline gpkit/build.py]: great inline comment

**Comment by galbramc** [comment]
Make sure you test the interactive stuff to make sure I didn't break it. I'm not sure how to run it or what it should do to make sure it still works correctly.

File: gpkit/small_scripts.py [hunk 2]
```diff
@@ -73,18 +60,10 @@ def is_sweepvar(sub):
         if sub[0] == "sweep":
             if isinstance(sub[1], Iterable) or hasattr(sub[1], "__call__"):
                 return True
-    except:
+    except:  # pylint: disable=bare-except
```
Comment (whoburg) [inline gpkit/small_scripts.py]: if really necessary, this could use an inline comment justifying, as in "great inline comment" above.

**Comment by whoburg** [comment]
review complete, this looks great. Only request was for an inline comment.

@bqpd, I agree with @galbramc that the warning related to `value_option` in `variables.py` seems potentially legit. Can you take a look at that? I know you fixed it (in the pylint sense) in another PR, so let's make sure this doesn't slip by when the other PR gets merged.

**Comment by whoburg** [comment]
controlpanel still works for a small model, which is a good sign for @galbramc's interactive fixes.

I'm ready to merge this once a) the inline comment is added, and b) the `value_option` issue is either resolved or captured in a ticket.

**Comment by whoburg** [comment]
other PRs will be full of pylint failures till we get this merged.

**Comment by galbramc** [comment]
I upgraded pylint and it found a couple more things to complain about.

Also, the mysterious duplicate code is now obvious:

```
gpkit/interactive/widgets.py:1: [R0801(duplicate-code), ] Similar lines in 2 files
==gpkit.interactive.sensitivity_map:46
==gpkit.nomials.nomial_core:77
            pvarstrs = ['%s^{%.2g}' % (varl, x) if "%.2g" % x != "1" else varl
                        for (varl, x) in pos_vars]
            nvarstrs = ['%s^{%.2g}' % (varl, -x)
                        if "%.2g" % -x != "1" else varl
                        for (varl, x) in neg_vars]
            pvarstrs.sort()
            nvarstrs.sort()
            pvarstr = ' '.join(pvarstrs)
            nvarstr = ' '.join(nvarstrs)
            c = mag(c)
            cstr = "%.2g" % c
            if pos_vars and (cstr == "1" or cstr == "-1"):
                cstr = cstr[:-1]
            else:
                cstr = latex_num(c)

            if not pos_vars and not neg_vars:
```
This begs for a separate function, and I don't think would be hard to fix. Your call, but it would suppress the last pylint warning that is likely to cause headaches.

**Comment by galbramc** [comment]
test this please

File: gpkit/small_scripts.py (hunk 2 above)
Comment (whoburg) [inline gpkit/small_scripts.py]: nevermind, I have a fix for this that I will push soon

**Comment by whoburg** [comment]
@galbramc, you pushed a bunch of pre-rebase commits that we don't want. I'm going to rebase and force push again, you'll have to delete your branch and check it out again.

File: gpkit/nomials/variables.py [hunk 3]
```diff
@@ -121,8 +121,10 @@ def __new__(cls, shape, *args, **descr):
                 descr["label"] = arg
 
         values = []
-        for value_option in ["value", "sp_init"]:
-            if value_option in descr:
+        value_option = None
+        for val_option in ["value", "sp_init"]:
+            if val_option in descr:
+                value_option = val_option
```
Comment (whoburg) [inline gpkit/nomials/variables.py]: I am dropping this commit -- this is now the second alternative proposed change to fix this issue (the first being 419ff866eda5e861266182c2177ef30e919442f8). Need @bqpd to review intent here before we disable the warning.

**Comment by whoburg** [comment]
force push incoming

**Comment by whoburg** [comment]
This will likely fail pylint tests due to the `value_option` warning in `variables.py` -- that's the only thing holding up merging -- @bqpd please review asap

**Comment by galbramc** [comment]
Ok, the duplicate code and value_option are the last two pylint warnings. I think they can easily be fixed before this is merged. (sorry about the rebase, I tried to do it right but forgot to fetch when the branch was deleted).

File: gpkit/build.py [hunk 4]
```diff
@@ -232,7 +233,8 @@ def look(self):
     def build(self):
         "Builds a dynamic library to GPKITBUILD or $HOME/.gpkit"
         try:
-            import ctypesgencore
+            # Testing the import, so the variable is intentionally not used
+            import ctypesgencore # pylint: disable=unused-variable
```
Comment (whoburg) [inline gpkit/build.py]: 👍

File: gpkit/nomials/variables.py [hunk 5]
```diff
@@ -148,6 +148,8 @@ def __new__(cls, shape, *args, **descr):
             i = it.multi_index
             it.iternext()
             descr.update({"idx": i})
+            assert value_option is not None
+            # TODO refactor logic for value_option and remove above assert
```
Comment (whoburg) [inline gpkit/nomials/variables.py]: for the record, this commit is by @whoburg; it's an amended version of 419ff866eda5e861266182c2177ef30e919442f8

**Comment by whoburg** [comment]
@galbramc, thanks for your enormously helpful effort on this.

I'm ready to merge this when tests pass. Let's let @bqpd review and press the button (and fix duplicate code issue, if inspired).

File: gpkit/nomials/variables.py (hunk 5 above)
Comment (galbramc) [inline gpkit/nomials/variables.py]: What value does value_option take on if the 'if' statement in the loop never returns true. Is the default now None or 'sp_init'?

**Comment by galbramc** [comment]
Let's not stop with just one little warning left! My vote is to fix the duplicate code issue and finally get to the magic number of 0 warnings! Don't forget that you started with about 2000 warnings. We are so close...

**Comment by galbramc** [comment]
@whoburg and @bqpd. You are now at ZERO pylint warnings. I don't know if you like my fix for the duplicate code, so feel free to change it or ditch it. But this now means you will never have to reset counts ever again (unless you decide you want to reintroduce pylint warnings...).

I am still a tad concerned with the value_option fix, see my inline comment.

File: gpkit/nomials/variables.py (hunk 5 above)
Comment (whoburg) [inline gpkit/nomials/variables.py]: if the `if value_option in descr` never returns true, then `value_option` takes the value `sp_init`. I agree this is likely not working as intended. @bqpd needs to review to determine the intent. I took his fix to the pylint for now.

**Comment by galbramc** [comment]
@whoburg, you forgot to update sensitivity_map with the latex helper function.

**Comment by whoburg** [comment]
yes I did, gonna amend and force push.

**Comment by whoburg** [comment]
Thanks @galbramc, that is a great fix for now, we need to refactor the latex code eventually, it's overcomplicated right now.

**Comment by galbramc** [comment]
Fair enough. What are your thoughts on my concerns with the value_option fix?

**Comment by whoburg** [comment]
#843

**Comment by galbramc** [comment]
Ok. I have no other objections now, feel free to merge as far as I am concerned.

**Comment by whoburg** [comment]
ok, @bqpd seems to be offline now; I'm going to merge this to make reviewing/merging other PRs less painful.

File: gpkit/nomials/variables.py (hunk 5 above)
Comment (bqpd) [inline gpkit/nomials/variables.py]: Why are you doing this in the for loop, though?

Comment (bqpd) [inline gpkit/nomials/variables.py]: @galbramc and @whoburg , by the construction of the code, `value_option` is guaranteed to have a value which makes sense if `values` is truthy.

**Comment by bqpd** [comment]
++

**Comment by bqpd** [comment]
woooot

File: gpkit/interactive/ractor.py [hunk 6]
```diff
@@ -106,7 +119,8 @@ def ractorjs(title, model, update_py, ranges, constraint_js=""):
                   varname = 'var'+i
                   idx = $w.ractive.data[varname]
                   console.log("$w-"+varname);
-                  document.getElementById("$w-"+varname).innerText = Math.round(100*$w.ranges[varname][idx])/100
+                  document.getElementById("$w-"+varname).innerText
```
Comment (bqpd) [inline gpkit/interactive/ractor.py]: Ooh hmm I think this will break the javascript.

Comment (galbramc) [inline gpkit/interactive/ractor.py]: can you test out the javascript? even better, write a unit tests that tests it out?

Comment (bqpd) [inline gpkit/interactive/ractor.py]: it's possible, but beyond scope for this script. If we have a full featured web interface we should do that.

Comment (whoburg) [inline gpkit/interactive/ractor.py]: does the possible breaking of javascript here need an issue?

Comment (bqpd) [inline gpkit/interactive/ractor.py]: Not for now, I don't think. Maybe when we have a full web framework that should become a goal. I fixed it on master.

**Pull request merged**